### PR TITLE
FIX: make sure scalarmappable updates are handled correctly in 3D

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

Code from a debugging session on the phone with @QuLogic 

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by QuLogic** [comment]
I don't think you want to rename attributes that way, as #18592 is going to do the opposite.

**Review by QuLogic — CHANGES_REQUESTED** [review]
Revert renames, due to the above comment. Also, unmark draft?

**Comment by tacaswell** [comment]
I did the renames because the 3D collections were inconsistent between each other.  I pushed to plural to match the 2D version, will push them the other way.

**Comment by tacaswell** [comment]
Because we have a shortage of people to review mplot3D, this being a critical bug, and in the interest of getting 3.3.3 out @QuLogic can merge this at his discretion on a single review.

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 1]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
         return np.min(vzs) if vzs.size else np.nan
 
 
+def _update_scalarmappable(sm):
+
+    """Update a 3D ScalarMappable.
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:646]:
```suggestion
    """
    Update a 3D ScalarMappable.
```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 2]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+    things that should be "in front" are rendered later to simulate
+    having a z-buffer (in addition to doing the projections).  This is
+    handled in the ``do_3d_projection`` methods which are called from the
+    draw method of the 3D Axes.  These methods:
+
+     - do the projection from 3D -> 2D
+     - internally sort based on depth
+     - stash the results of the above in the 2D analogs of state
+     - return the z-depth of the whole artist
+
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:664]:
```suggestion
    draw method of the 3D Axes.  These methods:

    - do the projection from 3D -> 2D
    - internally sort based on depth
    - stash the results of the above in the 2D analogs of state
    - return the z-depth of the whole artist

```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 3]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+    the last step is so that we can, at the Axes level, sort the children by
+    depth.
+
+    The base `draw` method of the 2D artists unconditionally call
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:668]:
```suggestion
    The base `draw` method of the 2D artists unconditionally calls
```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 4]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+    update_scalarmappable and rely on the method's internal caching logic to
+    lazily evaluate.
+
+    These things together mean you can have the sequence of events:
+
+    - we create the artist, to the color mapping and stash the results in a 3D
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:674]:
```suggestion
    - we create the artist, do the color mapping and stash the results in a 3D
```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 5]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+      specific state.
+    - change something about the ScalarMappable that marks it as in need of an
+      update (`ScalarMappable.changed` and friends).
+    - We call do_3d_projection and shuffle the stashed colors into the 2D version
+      of face colors
+    - the draw method calls the update_scalarmappable method which overwrites our
+      shuffled colors
+    - we get a render this is wrong
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:682]:
```suggestion
    - we get a render that is wrong
```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 6]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+    - if we re-render (either with a second save or implicitly via
+      tight_layout / constairned_layout / bbox_inches='tight' (ex via
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:684]:
```suggestion
      tight_layout / constrained_layout / bbox_inches='tight' (ex via
```

File: lib/mpl_toolkits/mplot3d/art3d.py [hunk 7]
```diff
@@ -635,6 +641,77 @@ def do_3d_projection(self, renderer=None):
...
+      inline's defaults)) we again shuffle the 3D colors
+    - because the CM is not marked as changed update_scalarmappable is a no-op and
+      we get a correct looking render.
+
+    This function is an internal helper to:
+
+     - sort out if we need to do the color mapping at all (has data!)
+     - sort out if update_scalarmappable is going to be a no-op
+     - copy the data over from the 2D -> 3D version
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/art3d.py:693]:
```suggestion
    This function is an internal helper to:

    - sort out if we need to do the color mapping at all (has data!)
    - sort out if update_scalarmappable is going to be a no-op
    - copy the data over from the 2D -> 3D version
```

File: lib/mpl_toolkits/tests/test_mplot3d.py [hunk 8]
```diff
@@ -1302,3 +1302,25 @@ def convert_lim(dmin, dmax):
...
+                                                 np.arange(0, 5, 1),
+                                                 np.arange(0, 5, 1)])))).T
+    c = x + y
+
+    # test
+    ax_test = fig_test.add_subplot(111, projection='3d')
+    sc_test = ax_test.scatter(x, y, z, c=c, s=40, cmap='viridis')
+    # force a draw
+    fig_test.canvas.draw()
+    # make it as "stale"
```
Comment (QuLogic) [inline lib/mpl_toolkits/tests/test_mplot3d.py:1321]:
```suggestion
    # mark it as "stale"
```

**Review by QuLogic — COMMENTED** [review]
Plus flake8 issues.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**